### ISSUE TITLE
M-FSDP: grad_norm deviates ~3.5e-5 from dist_opt/fsdp2 under gradient accumulation (num_microbatches>1)

### ISSUE THREAD

**Issue by conver334** [ISEEKYAN/Megatron-LM#149]
## Summary

Under gradient accumulation (`num_microbatches > 1`), the M-FSDP optimizer backend in `experimental/lite` produces a `grad_norm` that deviates from both reference backends by ~3.5e-5 relative, at **step 0**, with identical weights and a bit-identical loss. `dist_opt` and `fsdp2` agree with each other to ~2.7e-8 in the same run, so M-FSDP is the unique outlier.

The root cause appears to be that M-FSDP accumulates microbatch gradients in **bf16** (in the unsharded `full_param.grad`) and only converts to fp32 at the final reduce-scatter, whereas `dist_opt` and `fsdp2` accumulate in fp32.

`num_microbatches > 1` is the trigger.

## Environment

- `experimental/lite`, `examples/bench/correctness.py`, backend `mlite`
- Reproduced at `0d62b7f26`, `c1f7726a3`, and `fabba2d87` — identical numbers at all three
- torch 2.10.0+cu129, TE 2.12.0+5671fd3, 8x H100, single node
- Fixture: Qwen3-30B-A3B-Instruct-2507, `--truncate-layers 4 --keep-experts 2 --seq-len 128 --seed 42`
- Full determinism on: `MEGATRON_LITE_DETERMINISTIC=1`, `CUBLAS_WORKSPACE_CONFIG=:4096:8`, `NVTE_ALLOW_NONDETERMINISTIC_ALGO=0`, `CUDA_DEVICE_MAX_CONNECTIONS=8`

## Reproduction

```bash
cd experimental/lite
export PYTHONPATH=$PWD:/path/to/Megatron-LM
export MEGATRON_LITE_DETERMINISTIC=1 CUBLAS_WORKSPACE_CONFIG=:4096:8 NVTE_ALLOW_NONDETERMINISTIC_ALGO=0

for OPT in dist_opt fsdp2 mfsdp; do
  CUDA_DEVICE_MAX_CONNECTIONS=8 \
  torchrun --nnodes 1 --nproc_per_node 8 --master_port 29660 \
    examples/bench/correctness.py run --backend mlite \
    --hf-path <Qwen3-30B-A3B-Instruct-2507> --model-name qwen3_moe --impl lite \
    --steps 5 --seq-len 128 --seed 42 \
    --truncate-layers 4 --keep-experts 2 --same-data-across-dp \
    --override-optimizer-json '{"use_precision_aware_optimizer":false}' \
    --skip-weight-hash \
    --tp 1 --etp 1 --ep 1 --cp 1 --pp 1 --num-microbatches 2 \
    --impl-cfg-json "{\"optimizer\":\"${OPT}\"}" \
    --output-json ${OPT}.json
done
```

Then compare `steps[0].grad_norm.value` across the three JSONs. Step 0 is the diagnostic step: all three backends start from the identical checkpoint and report a bit-identical loss, so any `grad_norm` difference is purely a gradient-reduction artifact rather than trajectory drift.

## Observed — 2x2 ablation over (PP, num_microbatches)

Step-0 `grad_norm`, all backends from identical weights and bit-identical loss:

| config | PP | num_mb | dist_opt | fsdp2 | mfsdp | rel err dist_opt↔fsdp2 | rel err fsdp2↔mfsdp |
| --- | --- | --- | --- | --- | --- | --- | --- |
| c1 | 1 | 1 | 67.35098569 | 67.35099030 | 67.35098267 | 6.8e-08 | **1.1e-07** |
| a2 | 2 | 1 | 67.35098569 | 67.35099030 | 67.35098267 | 6.8e-08 | **1.1e-07** |
| a1 | 1 | 2 | 54.06295159 | 54.06295013 | 54.06108475 | 2.7e-08 | **3.5e-05** |
| c4 | 2 | 2 | 54.06294933 | 54.06295013 | 54.06108475 | 1.5e-08 | **3.5e-05** |

Reading the table:

- **PP is not the variable.** `a2` (PP=2, nmb=1) is bit-clean — M-FSDP matches `fsdp2` to 1.1e-07, exactly as at PP=1.
- **num_microbatches is the variable.** Both nmb=2 rows show a ~3.5e-05 relative deviation, ~300x the nmb=1 baseline.
- **The error is deterministic, not noise.** M-FSDP returns *exactly* `54.06108475` in both `a1` (PP=1) and `c4` (PP=2), and also in older runs at commits `c1f7726a3` and `0d62b7f26`. The two reference backends agree with each other to 2.7e-08 in the same runs.
- **The sign is consistent with precision loss:** M-FSDP's norm is systematically *lower* than the references, as expected from bf16 accumulation, not from a scaling error (a scaling bug would show a ratio near 2, not 1.00003).

By step 3–4 all three backends have drifted to ~1e-3 relative spread from ordinary chaotic divergence, which masks this signal — hence step 0 being the meaningful comparison.

## Suspected root cause

`megatron/lite/primitive/train_step.py:58` enables grad sync only on the last microbatch:

```python
if dist_opt and optimizer is not None and mb == num_microbatches - 1:
    optimizer.grad_sync_enabled = True
```

So for microbatches `0 .. n-2`, autograd accumulates into `spec.full_param.grad`, which is the grad of the all-gathered parameter in **compute dtype (bf16)**. The fp32 conversion only happens in `ParamBucket.prepare_grad_reduce` (`megatron/lite/primitive/optimizers/mfsdp/buffer.py:629-639`), which copies the already-accumulated bf16 grad into the fp32 `grad_comm_dtype` buffer just before the single reduce-scatter.

`MixedPrecisionPolicy` already defaults `main_grads_dtype` and `grad_comm_dtype` to fp32 (`mfsdp/config.py:40-42`) — the fp32 buffers exist, they are just not used as the *accumulator* across microbatches. `dist_opt` accumulates into an fp32 `main_grad` per microbatch, and `fsdp2` reduce-scatters per microbatch into an fp32 sharded `.grad`; both therefore avoid bf16 accumulation, which is consistent with those two agreeing to 2.7e-08.

## Impact

At 2 microbatches the error is ~3.5e-05, which passes any reasonable correctness gate — our ladder marks these configs PASS. But bf16 accumulation error grows with the number of microbatches, and real RL/SFT jobs use far more than 2. Since `grad_norm` feeds gradient clipping, a systematically under-reported norm means under-clipping relative to the reference backends.

## Suggested fix

Accumulate microbatch gradients in fp32 rather than in the bf16 `full_param.grad` — e.g. flush each microbatch's `full_param.grad` into the fp32 `main_grad_buffer` with `add_` instead of deferring all accumulation to bf16 and doing a single `copy_` at the end (`buffer.py:662` currently uses `copy_`, which also means the accumulate-into-fp32 path would need `add_` semantics plus a per-step zeroing).
